core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -2941,6 +2941,12 @@ private void registerQuery(
           SqlSelect.HAVING_OPERAND);
       registerSubQueries(selectScope2,
           SqlNonNullableAccessors.getSelectList(select));
+
+      if (enclosingNode.getKind() == SqlKind.UPDATE) {
```

Comment:
I wonder whether there are other statements that would need this.

Reply:
Thanks, at least MERGE suffers from the same issue.
Added the corresponding test and fix.